### PR TITLE
[Feat] support cp in sft trainer

### LINKED ISSUE #739: Add CP support
Currently, we cannot scale the training to long ctx (e.g. 64k) as the vram required to store activations is too large. Our only form of parallelism is FSDP which, being a DP dimension, does not allow scaling the micro batch size per gpu below 1. Supporting context parallelism (CP) and or tensor parallelism (TP) would allow us to split the computation for a sample between GPUs, allowing for micro batch size per gpu below 1.

Todo:
- [x] Add tp and cp to config
- [x] Create the correct device mesh in `model.py` given the passed config
- [x] `fully_shard` parameters correctly based on mesh
- [x] All models need to have an SDPA attention implementation to support CP (torch ring attn impl monkeypatches sdpa call)
- [ ] Loss needs to support local chunk computation 
  - [x] SFT
  - [ ] RL
- [x] Dataloader needs to support gpus loading the same data if they are part of the same CP/TP process group
- [x] Input ids need to be split by CP in the context_parallel context (is padding ok here?)
- [x] Add check for seqlen divisor

### LINKED ISSUE #740: Add TP support

### PR DESCRIPTION
**Pull request by Jackmin801**
Resolves https://github.com/PrimeIntellect-ai/prime-rl/issues/739

The MFU is horrible 7.7 vs 0.9 and there are random loss nans but it seems to be trending the right way 
image

```bash
uv run torchrun --nproc-per-node 8 src/prime_rl/trainer/sft/train.py \
  --model.name Jackmin108/glm-0.5B \
  --model.ac \
  --model.cp 1 \
  --data.pack-function cat \
  --model.trust_remote_code \
  --data.name PrimeIntellect/SFT-Math-Code-1K \
  --data.splits nemotron_math,nemotron_code \
  --data.seq-len 8192 \
  --data.batch-size 32 \
  --data.micro-batch-size 1 \
  --max-steps 1000 \
  --optim.lr 3e-4 \
  --scheduler.type cosine \
  --scheduler.warmup-steps 100 \
  --scheduler.min-lr 3e-7 \
  --ckpt.interval 100 \
  --output-dir ./outputs \
  --log.level debug \
  --monitor.wandb.project prime-rl-sft-debug
```

```bash
uv run torchrun --nproc-per-node 8 src/prime_rl/trainer/sft/train.py \
  --model.name Jackmin108/glm-0.5B \
  --model.ac \
  --model.cp 8 \
  --data.pack-function stack \
  --model.trust_remote_code \
  --data.name PrimeIntellect/SFT-Math-Code-1K \
  --data.splits nemotron_math,nemotron_code \
  --data.seq-len 8192 \
  --data.batch-size 32 \
  --data.micro-batch-size 1 \
  --max-steps 1000 \
  --optim.lr 3e-4 \
  --scheduler.type cosine \
  --scheduler.warmup-steps 100 \
  --scheduler.min-lr 3e-7 \
  --ckpt.interval 100 \
  --output-dir ./outputs \
  --log.level debug \
  --monitor.wandb.project prime-rl-sft-debug
```
---

**GitHub Issue**: https://github.com/PrimeIntellect-ai/prime-rl/issues/739
**Linear Issue**: Resolves [Issue ID]

### CODE REVIEW

**Comment by Jackmin801** [comment]
The MFU is horrible 7.7 vs 0.9 and there are random loss nans but it seems to be trending the right way 
image

```bash
uv run torchrun --nproc-per-node 8 src/prime_rl/trainer/sft/train.py \
  --model.name Jackmin108/glm-0.5B \
  --model.ac \
  --model.cp 1 \
  --data.pack-function cat \
  --model.trust_remote_code \
  --data.name PrimeIntellect/SFT-Math-Code-1K \
  --data.splits nemotron_math,nemotron_code \
  --data.seq-len 8192 \
  --data.batch-size 32 \
  --data.micro-batch-size 1 \
  --max-steps 1000 \
  --optim.lr 3e-4 \
  --scheduler.type cosine \
  --scheduler.warmup-steps 100 \
  --scheduler.min-lr 3e-7 \
  --ckpt.interval 100 \
  --output-dir ./outputs \
  --log.level debug \
  --monitor.wandb.project prime-rl-sft-debug
```

```bash
uv run torchrun --nproc-per-node 8 src/prime_rl/trainer/sft/train.py \
  --model.name Jackmin108/glm-0.5B \
  --model.ac \
  --model.cp 8 \
  --data.pack-function stack \
  --model.trust_remote_code \
  --data.name PrimeIntellect/SFT-Math-Code-1K \
  --data.splits nemotron_math,nemotron_code \
  --data.seq-len 8192 \
  --data.batch-size 32 \
  --data.micro-batch-size 1 \
  --max-steps 1000 \
  --optim.lr 3e-4 \
  --scheduler.type cosine \
  --scheduler.warmup-steps 100 \
  --scheduler.min-lr 3e-7 \
  --ckpt.interval 100 \
  --output-dir ./outputs \
  --log.level debug \
  --monitor.wandb.project prime-rl-sft-debug
```

**Review by samsja — APPROVED** [review]
lgtm

File: src/prime_rl/trainer/sft/data.py [hunk 1]
```diff
@@ -231,7 +244,91 @@ def __iter__(self) -> Iterator[Sample]:
...
 
-def collate(samples: list[Sample]) -> Batch:
+class StackDataset(StatefulIterableDataset):
+    """A dataset that stacks samples into batch with a fixed area"""
+
+    def __init__(self, dataset: StatefulIterableDataset, max_area: int):
+        self.dataset = dataset
+        self.max_area = max_area
+        self.current_seq_len = 0
+        assert math.log2(self.max_area).is_integer(), "max_area must be a power of 2"
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/sft/data.py:254]: bit manipulation is weird. no one will set max area of 1. denied

File: src/prime_rl/trainer/sft/data.py [hunk 2]
```diff
@@ -231,7 +244,91 @@ def __iter__(self) -> Iterator[Sample]:
...
+    def __iter__(self) -> Iterator[Sample]:
+        for sample in self.dataset:
+            # Add sample to packed samples
+            len_sample = len(sample["input_ids"])
+            if len_sample > self.max_area:
+                for key, value in sample.items():
+                    if key != "epoch":
+                        sample[key] = sample[key][: self.max_area]
+                len_sample = self.max_area
+            bucket_idx = int(math.log2(len_sample - 1)) + 1
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/sft/data.py:276]: no sample should have just bos token

File: src/prime_rl/trainer/sft/data.py [hunk 3]
```diff
@@ -231,7 +244,91 @@ def __iter__(self) -> Iterator[Sample]:
...
+            self.buckets[bucket_idx].append(sample)
+
+            if self.bucket_timers[bucket_idx] is not None:
+                hit_timeout = self.bucket_timers[bucket_idx] + self.bucket_timeout < self.step
+            else:
+                hit_timeout = False
+            if (2**bucket_idx) * len(self.buckets[bucket_idx]) >= self.max_area or hit_timeout:
+                if hit_timeout:
+                    while bucket_idx < len(self.buckets):
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/sft/data.py]: theres a bounds check above

File: src/prime_rl/trainer/sft/train.py [hunk 4]
```diff
@@ -146,38 +157,48 @@ def train(config: SFTTrainerConfig):
...
             assert input_ids.shape[0] == position_ids.shape[0]
 
-            # Forward pass
-            logits = forward(model, input_ids, position_ids).contiguous()
-            B, L, V = logits.shape
+            if config.model.cp > 1:
+                maybe_context_parallel = context_parallel(
+                    parallel_dims.world_mesh["cp"],
+                    buffers=tuple([input_ids, position_ids, target_ids, loss_mask]),
+                    buffer_seq_dims=(1, 1, 1, 1),
```
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/sft/train.py:164]: naa

File: src/prime_rl/trainer/sft/data.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Jackmin801) [reply src/prime_rl/trainer/sft/data.py]: i think u are right but also it shouldnt ever happen because its guaranteed to have broken before. i will just accept it without testing because im pretty sure its correct though it doesnt matter
